fix(Android): fix multiple padding related issues on Android (#5150, #5152, #5153, #5154)

shoamano83 opened the pull request:
<!--
PLEASE DON'T DELETE THIS TEMPLATE UNTIL YOU HAVE READ THE FIRST SECTION.

**What happens if you SKIP this step?**

Your pull request will NOT be evaluated!

PLEASE NOTE THAT PRs WITHOUT THE TEMPLATE IN PLACE WILL BE CLOSED RIGHT FROM THE START.

Thanks for helping us help you!
-->

### Does any other open PR do the same thing?

<!--
**Please keep in mind that we apply the FIFO rule for PRs, so if your PR comes after an existing one and there is no compelling reason to merge it instead of the existing one it will be discarded!**

If another PR exists that has similar scope to yours, please specify why you opened yours.
This could be one of the following (but not limited to)

 - the previous PR is stalled, as it's really old and the author didn't continue working on it
 - there are conflicts with the `master` branch and the author didn't fix them
 - the PR doesn't apply anymore (please specify why)
 - my PR is better (please specify why)
 -->

No

### What issue is this PR fixing?

- https://github.com/react-native-maps/react-native-maps/issues/5150
- https://github.com/react-native-maps/react-native-maps/issues/5152
- https://github.com/react-native-maps/react-native-maps/issues/5153
- https://github.com/react-native-maps/react-native-maps/issues/5154

### How did you test this PR?

<!--
Please let us know how you have verified that your changes work.

Ideally, your PR should contain a step-by-step test plan, so that reviewers can easily verify your changes.

Your PR will have much better chances of being merged if it is straightforward to verify.

Some questions you might want to think about:

- Which platform (eg. Android/iOS) & Maps API (eg. Google/Apple) does this change affect, if any?
- Did you test this on a real device, or in a simulator?
- Are there any platforms you were not able to test?
-->

I applied this change to my sample apps

- https://github.com/shoamano83/MapPaddingSample/tree/android_map_padding_on_map_ready
- https://github.com/shoamano83/MapPaddingSample/tree/android_initial_camera_and_map_padding
- https://github.com/shoamano83/MapPaddingSample/tree/android_map_padding_updated
- https://github.com/shoamano83/MapPaddingSample/tree/android_extra_padding

to confirm that padding now works on Android just same as iOS.

<!--
Thanks for your contribution :)
-->

shoamano83 commented on android/src/main/java/com/rnmaps/maps/MapManager.java:
```diff
@@ -193,7 +193,6 @@ public void setMapPadding(MapView view, @Nullable ReadableMap padding) {
         }
 
         view.applyBaseMapPadding(left, top, right, bottom);
-        view.map.setPadding(left, top, right, bottom);
```
`map.setPadding()` is already called during `applyBaseMapPadding()` so this line seems unnecessary.

shoamano83 commented:
I tried to split the change into multiple PRs but there were quite a few conflicting changes and it was difficult. Please allow me to consolidate the changes into a single PR.

It seems that `com.google.android.gms.maps.GoogleMap` doesn't reapply camera after `setPadding()` is invoked. (On iOS, it does.) This is the root cause of issues https://github.com/react-native-maps/react-native-maps/issues/5150, https://github.com/react-native-maps/react-native-maps/issues/5152 and https://github.com/react-native-maps/react-native-maps/issues/5153.

So,
- when we set up camera and padding, we should make sure to call `setPadding()` first, followed by `moveCamera()` / `animateCamera()`.
- when we update `mapPadding` property, we should reapply camera for the padding to work. (Otherwise, only Google logo moves.)

I also created a patch which can be applied by `patch-package`. Let me know if you have any feedback. Thanks!
[react-native-maps+1.18.0.patch](https://github.com/user-attachments/files/16791841/react-native-maps%2B1.18.0.patch)

shoamano83 commented:
It seems the fix for issue https://github.com/react-native-maps/react-native-maps/issues/5152 isn't perfect. Let me work on this PR a little more.

shoamano83 commented:
OK I fixed the issue and recreated this PR. It is review ready now.

Here's the patch file which can be applied by `patch-package`. It should also apply cleanly against v1.15.4 (for those who need to stick with older React Native release.)
[react-native-maps+1.18.0.patch](https://github.com/user-attachments/files/16817838/react-native-maps%2B1.18.0.patch)

shoamano83 commented:
Hi @salah-ghanim, this PR is review ready. May I ask for a review? Thanks!

salah-ghanim commented:
@shoamano83 thanks for the submission, I will soon start reviewing the bugs reported and the pull request

shoamano83 commented:
Hi @salah-ghanim, thank you for reviewing this PR!

My team uses this patch for a while and found an issue. Let me share it:
```
When `mapPadding` prop of a MapView is updated after a user drag the map, its camera is reset to `initialCamera`.

Steps to repro:
1. Set up a MapView with `initialCamera` prop
2. Drag the map to a different position
3. Update `mapPadding` prop of the MapView
```
The issue is caused by `lastCamera` introduced by this PR. I'm working on a fix.

shoamano83 commented:
> ```
> When `mapPadding` prop of a MapView is updated after a user drag the map, its camera is reset to `initialCamera`.
> 
> Steps to repro:
> 1. Set up a MapView with `initialCamera` prop
> 2. Drag the map to a different position
> 3. Update `mapPadding` prop of the MapView
> ```
> 
> The issue is caused by `lastCamera` introduced by this PR. I'm working on a fix.

I updated this PR to address this issue.

Highlights of commit f79f218:
- I reverted codes related to `lastCamera`.
- `applyBaseMapPadding()` includes extra logic to retrieve current camera (with base and edge paddings applied) and restore it after padding is updated.

I verified that with this change, issues #5150, #5152, #5153 and #5154 are still resolved.

Please kindly review again. Thank you!
